url/url.cpp

```diff
@@ -5,15 +5,65 @@
 #include "url/url.h"
 
 #include "util/string.h"
+#include "util/unicode.h"
 #include "util/uuid.h"
 
+#include <spdlog/spdlog.h>
+#include <unicode/bytestream.h>
+#include <unicode/idna.h>
+#include <unicode/putil.h>
+
 #include <array>
+#include <cassert>
+#include <charconv>
+#include <cmath>
 #include <cstdint>
+#include <cstdlib>
+#include <filesystem>
+#include <optional>
 #include <string>
+#include <string_view>
+#include <tuple>
+#include <utility>
 #include <variant>
+#include <vector>
 
 namespace url {
 
+void icu_init() {
+    static bool called_once = false;
+
+    if (called_once) {
```

Comment:
It can be called any number of times, it's just a waste of cycles to call it more than once.

tbh, the "correct" way to handle this would be to guarantee that this function runs once from a single thread before ICU gets used anywhere else (including outside of this library, if that makes sense). Is there a nice way to do that? Or do I have to lock it?

Reply:
We can do some static init trickery, but that's almost more gross than adding a lock/atomic since it happens outside of the normal program flow. The trickery:
```
// in file scope
struct IcuInitTrickery {
  IcuInitTrickery() { icu_init(); }
  ~IcuInitTrickery() { icu_kill(); }
};
IcuInitTrickery hack;
```